letsencrypt/models/letsencrypt.py

```diff
@@ -78,91 +78,108 @@ def validate_domain(self, domain):
-                '-key', self.generate_domain_key(domain),
-                '-subj', '/CN=%s' % domain, '-config', cfg.name,
-                '-out', csr,
-            ]
-            if len(domains) > 1:
-                cmdline.extend([
-                    '-reqexts', 'SAN',
-                ])
-            self.call_cmdline(cmdline)
-        return csr
+                    "or local domains!"))
 
     @api.model
     def cron(self):
+        parameter_model = self.env['ir.config_parameter']
         domain = urlparse.urlparse(
             self.env['ir.config_parameter'].get_param(
                 'web.base.url', 'localhost')).netloc
         self.validate_domain(domain)
-        account_key = self.generate_account_key()
-        csr = self.generate_csr(domain)
-        acme_challenge = get_challenge_dir()
-        if not os.path.isdir(acme_challenge):
-            os.makedirs(acme_challenge)
-        if self.env.context.get('letsencrypt_dry_run'):
-            crt_text = 'I\'m a test text'
-        else:  # pragma: no cover
-            from acme_tiny import get_crt, DEFAULT_CA
-            crt_text = get_crt(
-                account_key, csr, acme_challenge, log=_logger, CA=DEFAULT_CA)
-        with open(os.path.join(get_data_dir(), '%s.crt' % domain), 'w')\
-                as crt:
-            crt.write(crt_text)
-            chain_cert = urllib2.urlopen(
-                self.env['ir.config_parameter'].get_param(
-                    'letsencrypt.chain_certificate_address',
-                    'https://letsencrypt.org/certs/'
-                    'lets-encrypt-x3-cross-signed.pem')
-            )
-            crt.write(chain_cert.read())
-            chain_cert.close()
-            _logger.info('wrote %s', crt.name)
-        reload_cmd = self.env['ir.config_parameter'].get_param(
-            'letsencrypt.reload_command', False)
-        if reload_cmd:
-            _logger.info('reloading webserver...')
-            self.call_cmdline(['sh', '-c', reload_cmd])
+        # Generate account key
+        account_key_file = self._generate_key('account_key')
+        account_key = jose.JWKRSA.load(open(account_key_file).read())
+        # Generate domain key
+        domain_key_file = self._generate_key(domain)
+        net = client.ClientNetwork(account_key)
+        if self.env.context.get('DRY_RUN'):
+            directory_url = V2_STAGING_DIRECTORY_URL
+        else:
+            directory_url = V2_DIRECTORY_URL
+        directory_json = requests.get(directory_url).json()
+        acme = client.ClientV2(directory_json, net)
+        reg = Registration(
+            key=account_key.public_key(),
+            terms_of_service_agreed=True)
+        new_reg = NewRegistration(**dict(reg))
+        acme.new_account(new_reg)
+        altnames = parameter_model.search(
+            [('key', 'like', 'letsencrypt.altname.')],
+            order='key')
+        valid_domains = [x.value for x in altnames] + [domain]
```

Comment:
I think this should be the other way around. The first entry will become the CN of the certificate, right?